### ISSUE TITLE
Breakage to `jiff` due to `ambiguous_negative_literals`

### ISSUE THREAD

**Issue by traviscross** [rust-lang/rust#128287]
We should discuss this bit of evidence that came in after the FCP completed on:

- https://github.com/rust-lang/rust/pull/121364

@BurntSushi [said](https://github.com/rust-lang/rust/issues/121364#issuecomment-2254136058):

> This completely broke Jiff once the change landed. Because it specifically supports negating spans like `-1.hour()`. [In the case of Jiff, there is no ambiguity because `-1.hour()` and `(-1).hour()` and `-(1.hour())` are all precisely equivalent.](https://docs.rs/jiff/latest/jiff/struct.Span.html#negative-spans)

cc @BurntSushi @rust-lang/lang

@rustbot labels +T-lang +I-lang-nominated

**Comment by workingjubilee** [comment]
This lint indeed assumes the operations are not associative.

**Comment by Urgau** [comment]
The partially-uplifted Clippy `clippy::precedence` lint (which made up this lint) had a [basic heuristic](https://github.com/rust-lang/rust/blob/6ae8003ff55971fc3dfb4e6197aaad89affb4876/src/tools/clippy/clippy_lints/src/precedence.rs#L10-L25) for not linting in cases where the precedence of the operations didn't matter. It was part of the original PR, but `@RalfJung` preferred consistent linting [[comment]](https://github.com/rust-lang/rust/pull/117161#discussion_r1371859160):

> Personally I'd prefer it to warn consistently, even if the semantics of the function mean it doesn't matter. (a) that seems more consistent and less surprising, and (b) I am writing tests which ensure that `sin` behaves correctly, and I care about which sign the argument actually has.

**Comment by Urgau** [comment]
If it's decided that it's not worth linting on these cases, I would propose using the `#[diagnostic]` namespace by introducing an opt-out, that library authors could use on to suppress the lint for downstream users. Something like `#[diagnostic::irrelevant_negative_literal_precedence]` *(better name to be found)* which could applied to all the relevant function definitions.

**Comment by traviscross** [comment]
We discussed this in our lang triage call today.

@Urgau: If you would, we'd like this lint changed to `allow-by-default` while we consider this question further.

cc @compiler-errors

**Comment by Urgau** [comment]
As asked the lint was changed to `allow-by-default` in https://github.com/rust-lang/rust/pull/128449.

**Comment by traviscross** [comment]
@BurntSushi: What do you think of the proposal that @Urgau makes [here](https://github.com/rust-lang/rust/issues/128287#issuecomment-2254580843), as it applies to `jiff`.

I'd also be interested to hear your thoughts on the question of whether something like this should be opt-in or opt-out?

**Comment by BurntSushi** [comment]
@traviscross I love @Urgau's idea. Will it be possible to use that annotation on older Rust compilers where it will just be ignored? Otherwise it would be hard for me to use it until my MSRV matches the Rust release where the annotation becomes available. (Or do a build script with conditional compilation.)

In terms of opt-in or opt-out, I agree that opt-out makes the most sense. Otherwise I think this lint would probably not be nearly as effective.

**Comment by compiler-errors** [comment]
Using an attribute from the diagnostic namespace -- even an unknown one -- will bump your MSRV to 1.78, when the diagnostic namespace was stabilized. Before that you'll get resolver errors that afaict you cannot suppress.

**Comment by BurntSushi** [comment]
@compiler-errors I think that should be fine. Probably by the time everything lands on stable, I'll soon thereafter be able to bump my MSRV to 1.78. Worst case, I'll bite the bullet and do version sniffing in `build.rs`.

**Comment by compiler-errors** [comment]
Yep, and 1.78 should be the MSRV for all future tricks of this same sort.

**Comment by RodBurman** [comment]
I wrote a piece of test code for this:

```
use jiff::{Span, ToSpan};

#[warn(ambiguous_negative_literals)]
fn main() {
    let span : Span = -5i32.hours();   // lint should fire
    let spn2 : Span = (-5i32).hours();
    let neg1 = -1i32.abs();  // lint should fire
    let abs1 = (-1i32).abs();
    println!("span '{span:?}' spn2 '{spn2:?}' neg1 {neg1:?} abs1 {abs1:?}");
}
```

Which compiles with the expected warnings:
```
% cargo build   
   Compiling jiffclip v0.1.0 (/Users/rod/code/rust/triage/2026/06/jiffclip)
warning: `-` has lower precedence than method calls, which might be unexpected
 --> src/main.rs:5:23
  |
5 |     let span : Span = -5i32.hours();
  |                       ^^^^^^^^^^^^^
  |
  = note: e.g. `-4.abs()` equals `-4`; while `(-4).abs()` equals `4`
note: the lint level is defined here
 --> src/main.rs:3:8
  |
3 | #[warn(ambiguous_negative_literals)]
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: add parentheses around the `-` and the literal to call the method on a negative literal
  |
5 |     let span : Span = (-5i32).hours();
  |                       +     +
help: add parentheses around the literal and the method call to keep the current behavior
  |
5 |     let span : Span = -(5i32.hours());
  |                        +            +

warning: `-` has lower precedence than method calls, which might be unexpected
 --> src/main.rs:7:16
  |
7 |     let neg1 = -1i32.abs();
  |                ^^^^^^^^^^^
  |
  = note: e.g. `-4.abs()` equals `-4`; while `(-4).abs()` equals `4`
help: add parentheses around the `-` and the literal to call the method on a negative literal
  |
7 |     let neg1 = (-1i32).abs();
  |                +     +
help: add parentheses around the literal and the method call to keep the current behavior
  |
7 |     let neg1 = -(1i32.abs());
  |                 +          +

warning: `jiffclip` (bin "jiffclip") generated 2 warnings
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.76s
```

and has the expected output:

> % ./target/debug/jiffclip 
> span '5h ago' spn2 '5h ago' neg1 -1 abs1 1

I am not sure what the issue is here. Yes span and spn2 produces the same results '5h ago' but given that i32 (and i64, etc.) are much more common in code than "jiff" I would find the lint much more useful as a default. A new (or tired) rust programmer would almost certainly think -1.abs() is 1 and so being warned that it is in fact -1 would be helpful, then having learned (or had a refreshing cup of tea/coffee) would know to write the second form and subsequently write it many many times, so when the time came to use jiff then "(-5).hours()" would come naturally by analogy and so the lint would not fire.

I am basically think the lint would be more useful if on by default than any harm it would cause those using jiff.

**Comment by BurntSushi** [comment]
@RodBurman I think we have a solution here that makes everyone happy: https://github.com/rust-lang/rust/issues/128287#issuecomment-2254580843
